[Lens] [ES|QL] Lens ES|QL flyout "Apply and close" button's UX improvement

awahab07 opened the pull request:
Fixes https://github.com/elastic/kibana/issues/243080
Fixes https://github.com/elastic/kibana/issues/243770

## Summary

To address the UX quirks mentioned the the parent issues, the PR:
- Integrates syntax validation in Lens ES|QL editor to disable "Apply and close" button on Lens ES|QL panel configuration flyout when syntax errors are present
- Shows contextual tooltip messages explaining why the button is disabled ~Shows the error help text if updated query hasn't been run, while disabling "Apply and close" button.~


### ES|QL flyout form validation behavior

The "Apply and close" button is now disabled when:
- ~ES|QL query has syntax errors (detected client-side)~ (Now, query needs to be run first in order to let flyout know if there are syntax errors).
- ES|QL query has been changed but not yet run
- ES|QL query has errors after execution


### Video

The following video demonstrates how the fix addresses the issues mentioned in the parent issues:

https://github.com/user-attachments/assets/8396425b-245d-44bc-8acb-1cc721aeac3b

**Update 01**: [Help text instead of tooltips](https://github.com/elastic/kibana/pull/247625#issuecomment-3710798464).

**Update 02**: [No help text either](https://github.com/elastic/kibana/pull/247625#issuecomment-3716626165).

**Update 03**: [Only a single tooltip](https://github.com/elastic/kibana/pull/247625#issuecomment-3746000054).

awahab07 committed:
Disable Apply button for ES|QL validation errors

- Add client-side ES|QL syntax error detection using Parser.parseErrors()
- Disable 'Apply and close' button when:
  - ES|QL query has syntax errors
  - ES|QL query has been changed but not yet run
  - ES|QL query has runtime errors after execution
- Show contextual tooltip messages explaining why button is disabled
- Remove isNewPanel bypass so new panels are also validated
- Add getESQLQuerySyntaxErrors() utility to kbn-esql-utils package

awahab07 committed:
Merge remote-tracking branch 'origin/main' into 243080_243770_ESQL_Lens-Apply-and-clse-button-UX-Improvements

elasticmachine commented:
Pinging @elastic/kibana-visualizations (Team:Visualizations)

awahab07 committed:
Handle thrown errors in getESQLQuerySyntaxErrors. Highlight the varied behavior via test.

awahab07 committed:
Remove backtick test.

awahab07 committed:
Merge remote-tracking branch 'origin/main' into 243080_243770_ESQL_Lens-Apply-and-clse-button-UX-Improvements

kibanamachine committed:
Changes from node scripts/notice

awahab07 committed:
Simplify UX, show help text instead when query hasn't been run.

awahab07 committed:
Merge remote-tracking branch 'origin/main' into 243080_243770_ESQL_Lens-Apply-and-clse-button-UX-Improvements

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/377910)
* Commit: d6bf9948ef3887e6bc04f10351bff3d2d04fa8e8

### Failed CI Steps
* [x-pack/solutions/security/test/cloud_security_posture_functional/config.ts](https://buildkite.com/elastic/kibana-pull-request/builds/377910#019b8ea2-147c-4761-9ea4-708cb289cc79)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/377910#019b8ea2-147c-4761-9ea4-708cb289cc79) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/377910/jobs/019b8ea2-147c-4761-9ea4-708cb289cc79/artifacts/019b8ec9-f711-454e-b9ab-e02f3c5e1d88) x-pack/solutions/security/test/cloud_security_posture_functional/config.ts / Cloud Security Posture Security Network Page - Graph visualization ECS fields only expanded flyout - entity enrichment for multiple actors and targets

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/47059dc1b5fb6c18adde9f80fcad3939a1e8f861) | [after](https://github.com/elastic/kibana/commit/d6bf9948ef3887e6bc04f10351bff3d2d04fa8e8) | diff |
| --- | --- | --- | --- |
| `@kbn/esql-utils` | 126 | 128 | +2 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/47059dc1b5fb6c18adde9f80fcad3939a1e8f861) | [after](https://github.com/elastic/kibana/commit/d6bf9948ef3887e6bc04f10351bff3d2d04fa8e8) | diff |
| --- | --- | --- | --- |
| `lens` | 1.9MB | 1.9MB | +943.0B |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/47059dc1b5fb6c18adde9f80fcad3939a1e8f861) | [after](https://github.com/elastic/kibana/commit/d6bf9948ef3887e6bc04f10351bff3d2d04fa8e8) | diff |
| --- | --- | --- | --- |
| `lens` | 64.6KB | 64.8KB | +228.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/47059dc1b5fb6c18adde9f80fcad3939a1e8f861) | [after](https://github.com/elastic/kibana/commit/d6bf9948ef3887e6bc04f10351bff3d2d04fa8e8) | diff |
| --- | --- | --- | --- |
| `@kbn/esql-utils` | 179 | 181 | +2 |


</details>


### History
* :broken_heart: [Build #377434](https://buildkite.com/elastic/kibana-pull-request/builds/377434) failed de95b04c8b198e5106e6ff608158207f3da571b5
* :broken_heart: [Build #377433](https://buildkite.com/elastic/kibana-pull-request/builds/377433) failed db451cc87a05f83bf75587904109873fd25c3c76
* :broken_heart: [Build #377178](https://buildkite.com/elastic/kibana-pull-request/builds/377178) failed 2c1f5d8398d2c581dcb738f6dc45ac5560913132

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/377910","number":377910,"commit":"d6bf9948ef3887e6bc04f10351bff3d2d04fa8e8"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/377434","number":377434,"commit":"de95b04c8b198e5106e6ff608158207f3da571b5"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/377433","number":377433,"commit":"db451cc87a05f83bf75587904109873fd25c3c76"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/377178","number":377178,"commit":"2c1f5d8398d2c581dcb738f6dc45ac5560913132"}],"number":377910}
buildkite-pr-comment-->

awahab07 committed:
Merge `origin/main` into current branch.

awahab07 committed:
Simplify Apply button validation for ES|QL queries

- Disable Apply button when query is pending submission or has errors
- Remove isNewPanel bypass so new panels are also validated
- Remove getESQLQuerySyntaxErrors (client-side syntax validation)
- Remove help text and tooltip from Apply button
- Button is simply disabled without additional messaging

awahab07 committed:
Restore query_parsing_helpers.test.ts from main to avoid whitespace changes

awahab07 committed:
Merge remote-tracking branch 'origin/main' into 243080_243770_ESQL_Lens-Apply-and-clse-button-UX-Improvements

# Conflicts:
#	x-pack/platform/plugins/shared/lens/public/app_plugin/shared/edit_on_the_fly/lens_configuration_flyout.tsx

teresaalvarezsoler reviewed (approved):
Works very well! If we can add the tooltip when the button is disabled, that would be great.

awahab07 committed:
Merge remote-tracking branch 'origin/main' into 243080_243770_ESQL_Lens-Apply-and-clse-button-UX-Improvements

awahab07 committed:
"Run ES|QL query ..." tooltip on 'Apply and close' button when button is disabled due to pending query.

awahab07 committed:
Fix issue where changing the chart type from Treemap to Pie on a saved ES|QL Discover Session would not enable "Apply and close" button and apply the changed suggestion even when user clicks "Cancel" button.

awahab07 commented:
The tooltip on the "Apply and close" button when the button is disabled due to an unrun/pending ES|QL query, has been [added](https://github.com/elastic/kibana/pull/247625/commits/221c92575d90f032f84b0f1cc392fe06d6bd68a0) back.


https://github.com/user-attachments/assets/9163910c-8181-4e71-bc00-26da4dbc4385

awahab07 commented:
Note that the issue mentioned in [comment above](https://github.com/elastic/kibana/pull/247625#issuecomment-3723298464) has been [fixed](https://github.com/elastic/kibana/pull/247625/commits/c267cb4f93ab747fac8da34e016cc94980f01b4c) in the same PR as it was making an e2e test fail.

The bug:

https://github.com/user-attachments/assets/3b172af5-c78e-4f42-b2e4-e77d9ad0ddd5

After the fix:

https://github.com/user-attachments/assets/c1b33db4-178a-4498-b716-eec231dcaae7

awahab07 commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'main' into 243080_243770_ESQL_Lens-Apply-and-clse-button-UX-Improvements

stratoula reviewed (approved):
This looks exactly what I had in mind and is not very complex. I havent checked the code but from the functionality perspective this seems right! Thank you a lot for applying my feedback

jughosta reviewed (approved):
Data Discovery changes LGTM

markov00 reviewed (changes requested):
Please just wait for our eng review

elasticmachine committed:
Merge branch 'main' into 243080_243770_ESQL_Lens-Apply-and-clse-button-UX-Improvements

elasticmachine committed:
Merge branch 'main' into 243080_243770_ESQL_Lens-Apply-and-clse-button-UX-Improvements

elasticmachine committed:
Merge branch 'main' into 243080_243770_ESQL_Lens-Apply-and-clse-button-UX-Improvements

markov00 reviewed (approved):
Tested locally, LGTM

awahab07 committed:
Merge `origin/main` into current branch.

awahab07 committed:
Fix import

awahab07 committed:
Merge remote-tracking branch 'origin/main' into 243080_243770_ESQL_Lens-Apply-and-clse-button-UX-Improvements
